Dataplane policy states and events API swagger

NarineM opened the pull request:
### Latest improvements:
Added data plane events and states API under version 2019-10-01. Added samples. 
<br>

@pilor

NarineM committed:
swagger documentation for data plane PolicyStates API

NarineM committed:
enable events api in 2019-10-01 api version

NarineM committed:
data plane events API and data plane states API fixes

NarineM committed:
Data plane sample fixes and resolving autorest tool compile errors

Data plane sample fixes and resolving autorest tool compile errors

NarineM committed:
Added more examples for dp and minor fixes

Added more examples for dp and minor fixes

NarineM committed:
Update api version in exisiting events API examples

Update api version in exisiting events API examples to be 2019-10-01

AutorestCI commented:
/azp run automation - sdk

azuresdkci commented:
Can one of the admins verify this patch?

NarineM committed:
Fix model validator errors

Fix model validator errors

AutorestCI commented:
/azp run automation - sdk

NarineM committed:
Remove aggrigate data plane events and states examples from swagger

Remove aggrigate data plane events and states examples from swagger

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

ravbhatnagar commented:
@NarineM - can you please set something up with armapireview to go over this? It will be much quicker in person since this is a new type.

NarineM commented:
@ravbhatnagar  ok, thanks, I will setup a meeting.

NarineM committed:
Revert - Add back references to aggrigated component counts

Revert - Add back references to aggrigated component counts

AutorestCI commented:
/azp run automation - sdk

raych1 commented on specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/examples/PolicyEvents_QueryResourceScopeExpandComponentsGroupByWithAggregate.json:
```diff
@@ -0,0 +1,59 @@
+                    "tenantId": "72f988bf-86f1-41af-91ab-2d7cd011db47",
+                    "principalOid": "0d81b461-6bb0-4909-a102-d51803a7d275",
+                    "policyDefinitionAction": "audit",
+                    "totalActions": 6
```
Property of `totalActions` is not defined in the swagger [here](https://github.com/Azure/azure-rest-api-specs/blob/77c82d458cdef45277b91b4351075383e0eee776/specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json#L755)

NarineM replied:
The issue must be something else, because there are quite a few other json files that are written in the same manner, where they take custom property names from the  request URL and are not defined in the schema, for some reason model validation is passing for these root properties but not for nested properties(that are in my PR). Here is an example of similar json for which validation is passing:
https://github.com/Azure/azure-rest-api-specs/blob/77c82d458cdef45277b91b4351075383e0eee776/specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/examples/PolicyEvents_FilterAndAggregateOnly.json

If you look in this example NumDenyEvents property is not defined in the swagger for this API, but validation is passing.

raych1 commented:
@NarineM , there're two additional property not defined in swagger which causes the model validation failure. One I have already commented inline. another one is  in `specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/examples/PolicyStates_QueryResourceScopeExpandComponentsGroupByWithAggregate.json   jsonPath: '$.value[0].components.components[0].count'` . Please check the build log for details.

NarineM commented:
@raych1, the root cause is something else please see this json for which validation is passing where it contains a property NumDenyEvents  which is not defined in events API swagger at: https://github.com/NarineM/azure-rest-api-specs/blob/narinem-dataplane-swagger/specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json

Here is the json with similar implementation but Model validation is passing for this one:
{
  "parameters": {
    "policyEventsResource": "default",
    "subscriptionId": "fffedd8f-ffff-fffd-fffd-fffed2f84852",
    "api-version": "2019-10-01",
    "$from": "2018-02-05T18:00:00Z",
    "$filter": "PolicyDefinitionAction eq 'deny'",
    "$apply": "aggregate($count as NumDenyEvents)"
  },
  "responses": {
    "200": {
      "body": {
        "@odata.context": "https://management.azure.com/subscriptions/fffedd8f-ffff-fffd-fffd-fffed2f84852/providers/Microsoft.PolicyInsights/policyEvents/$metadata#default",
        "@odata.count": 1,
        "value": [
          {
            "@odata.id": null,
            "@odata.context": "https://management.azure.com/subscriptions/fffedd8f-ffff-fffd-fffd-fffed2f84852/providers/Microsoft.PolicyInsights/policyEvents/$metadata#default/$entity",
            "NumDenyEvents": 40
          }
        ]
      }
    }
  }
}

raych1 commented:
Thanks @NarineM for raising this. I will double check.

raych1 commented:
@NarineM , you need to set `additionalProperties` in the model if it will have undefined properties. Please see [example](https://github.com/Azure/azure-rest-api-specs/blob/77c82d458cdef45277b91b4351075383e0eee776/specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json#L748)

NarineM committed:
Added additional properties bag to data plane json object

Added additional properties bag to data plane json object

AutorestCI commented:
/azp run automation - sdk

NarineM committed:
Added policyEvents for 2019-10-01 in readme

AutorestCI commented:
/azp run automation - sdk

NarineM committed:
fixed sample ref file path

AutorestCI commented:
/azp run automation - sdk

NarineM commented:
@raych1 thanks that worked. Now I am getting breaking change violation on "components" property being published under same API version. Here is the error from validating breaking change : https://dev.azure.com/azure-sdk/public/_build/results?buildId=303511&view=logs&j=08b5038d-1e84-5877-88da-859df8312b69&t=c1978ecd-fc8a-5291-6d7a-7dd459d8b50f

If I need this change be published for the same API , how do I bypass the breaking change validation error?

raych1 commented:
@NarineM , prettier failure is about JSON format issue. Please follow https://aka.ms/AA6h31t to fix the issue. 
Introduce new property in stable version API is considered as breaking change. Can that be published with new version? If not, it needs to request approval on force merge.

NarineM commented:
@raych1 I have ran Prettier check as it is documented at the link you provided  , but can not figure out what is wrong. Looking at the log looks like the tool is not even running. This is what I get when I run the tool:
PS C:\git\azure-rest-api-specs> npm run prettier -- --write "specification/policyinsights/**/*.json"

> azure-rest-api-specs-tests@0.1.0 prettier C:\git\azure-rest-api-specs
> prettier "--write" "specification/policyinsights/**/*.json"

[error] Unable to expand glob patterns: specification/policyinsights/**/*.json !**/node_modules/** !./node_modules/** !**/.{git,svn,hg}/** !./.{git,svn,hg}/**
[error] Cannot find module 'C:\git\azure-rest-api-specs\scripts\prettier-swagger-plugin' from 'C:\git\azure-rest-api-specs'
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! azure-rest-api-specs-tests@0.1.0 prettier: `prettier "--write" "specification/policyinsights/**/*.json"`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the azure-rest-api-specs-tests@0.1.0 prettier script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\narinem\AppData\Roaming\npm-cache\_logs\2020-03-17T18_49_42_976Z-debug.log

NarineM commented:
@raych1 I have validated that these json files are correct based on RFC 8259. Using tools like https://jsonformatter.curiousconcept.com/
so I am not quite sure whether it is an issue with the tool or the files?

anthony-c-martin commented:
I'm happy from an ARM perspective as we're really just adding filtering options to an existing API. @pilor is going to review from a Policy perspective.

raych1 commented:
> @raych1 I have validated that these json files are correct based on RFC 8259. Using tools like https://jsonformatter.curiousconcept.com/
> so I am not quite sure whether it is an issue with the tool or the files?

The tool only detects the style issue in JSON. Can you make sure the script `C:\git\azure-rest-api-specs\scripts\prettier-swagger-plugin` exist locally?

pilor commented on specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyStates.json:
```diff
@@ -1227,6 +1233,13 @@
+            }
+        },
+        "components": {
+          "description": "Components state compliance records. By default this is not populated.",
```
>By default this is not populated. [](start = 63, length = 33)

Instead of saying this can you specify when this would be populated?

NarineM replied:
fixed.

pilor commented on specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json:
```diff
@@ -0,0 +1,1024 @@
+          "type": "string"
+        },
+        "components": {
+          "description": "Components event records. By default this is not populated.",
```
>By default this is not populated. [](start = 52, length = 33)

Same comment as policyStates.  Please specify when this is populated

NarineM replied:
Sure, fixing it now.

pilor commented on specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json:
```diff
@@ -0,0 +1,1024 @@
+          "type": "string",
+		  "format": "date-time"
+        },
+		"tenantId": {
```
>"tenantId" [](start = 2, length = 10)

This can be different than the root policyEvent tenantId?

NarineM replied:
Yes, the user for keyvault can be different from the user operating over certificates that belong to given keyvault.

pilor commented on specification/policyinsights/resource-manager/Microsoft.PolicyInsights/stable/2019-10-01/policyEvents.json:
```diff
@@ -0,0 +1,1024 @@
+      "in": "query",
+      "required": false,
+      "type": "string",
+      "description": "The $expand query parameter. For example, to expand policyEvaluationDetails, use $expand=policyEvaluationDetails or to expand Components use $expand=components",
```
> to expand policyEvaluationDetails, use $expand=policyEvaluationDetails [](start = 63, length = 71)

policyEvaluationDetails doesn't exist in policyEvents right?

NarineM replied:
Ah, yes, thanks for catching it. Fixed.

pilor reviewed (changes requested):
A few minor comments

NarineM commented:
@raych1 yes, the swagger prettier script exists in the path, I have pulled the entire repo that sits on github.

NarineM committed:
Addressed PR review feedback

Addressed PR review feedback

AutorestCI commented:
/azp run automation - sdk

NarineM commented:
@pilor Just pushed a new iteration that addresses your comments. I can see this warning from cli-extension, do you know what it complains about?:
"azure-cli-extensions
No readme.md specification configuration files were found that are associated with the files modified in this pull request, or swagger_to_sdk section in readme.md is not configured "

pilor commented:
> @pilor Just pushed a new iteration that addresses your comments. I can see this warning from cli-extension, do you know what it complains about?:
> "azure-cli-extensions
> No readme.md specification configuration files were found that are associated with the files modified in this pull request, or swagger_to_sdk section in readme.md is not configured "

No, I don't know the specifics on that but I have seen it on my own PRs which went through just fine.  The SDK reviewer is a better contact.

NarineM committed:
Fix SDK automation error

SDK automation is throwing an error if the parameter descriptions for $expand do not match across PolicyStates and PolicyEvents API.

AutorestCI commented:
/azp run automation - sdk

NarineM commented:
@raych1 Can you help with this new error from SDK automation log?, it says branch changes need to be squished. Looks like something with the tool.

raych1 commented:
@NullMDR , can you help with the SDK automation failures?

NullMDR commented:
/azp run

Commit:
Fix prettier

NarineM commented:
@raych1  thanks. So far all PR merge validation issues have been resolved except prettier check. This PR can be merged unless you tell me the prettier check issue is blocking. As of now I can not figure out why the tool is not running or what it complains about.

NarineM commented:
Please go ahead with merging this PR when you get a chance if you do not find anything with prettier check.

raych1 commented:
@NarineM, prettierCheck is required to pass. @NullMDR , can you help Narine, he failed to run prettier?

AutorestCI commented:
/azp run automation - sdk

NullMDR commented:
@NarineM I've fixed your prettier failure by running 
`
npm run prettier -- --write "specification/policyinsights/**/*.json"
`
But please fix the merge conflict.

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

NarineM committed:
Merge branch 'master_upstream' into narinem-dataplane-swagger

AutorestCI commented:
/azp run automation - sdk

NarineM commented:
@NullMDR, thank you. I have resolved merge conflics. The only failure now remaining is related to api version (under breaking change catergory).  As we discussed earlier this change needs to be published for the same api version 2019-10-01. Please go ahead with merging this PR.

NullMDR commented:
@jianyexi Why adding property in response will be considered breaking change?

NarineM commented:
@NullMDR  It is not breaking change because this property is only appended when the client specifies it in request URL with $expand clause. This functionality is available in the current version of API on the RP side, so it should not be published under new API version.

NarineM commented:
@NullMDR , can you go ahead with merging this PR? Let me know what is blocking if any. thanks.

ArcturusZhang commented:
> @NullMDR It is not breaking change because this property is only appended when the client specifies it in request URL with $expand clause. This functionality is available in the current version of API on the RP side, so it should be published under the same API version 2019-10-01 and not a new version.

I think this is reasonable. 

Hi @akning-ms would you please have us merge this PR? This breaking change should be acceptable based on the author's comments
